Allocation Profiler: Types for all allocations

NHDaly opened the pull request:
A variant on https://github.com/JuliaLang/julia/pull/50333.

Pass the types to the allocator functions.

-------

Before this PR, we were missing the types for allocations in two cases:

1. allocations from codegen
2. allocations in `gc_managed_realloc_`

The second one is easy: those are always used for buffers, right?

For the first one: we extend the allocation functions called from codegen, to take the type as a parameter, and set the tag there.

I kept the old interfaces around, since I think that they cannot be removed due to supporting legacy code?

------

An example of the generated code:
```julia
  %ptls_field6 = getelementptr inbounds {}**, {}*** %4, i64 2
  %13 = bitcast {}*** %ptls_field6 to i8**
  %ptls_load78 = load i8*, i8** %13, align 8
  %box = call noalias nonnull dereferenceable(32) {}* @ijl_gc_pool_alloc_typed(i8* %ptls_load78, i32 1184, i32 32, i64 4366152144) #7
```

Fixes #43688.
Fixes #45268.


<img width="887" alt="Screenshot 2023-07-29 at 8 48 16 PM" src="https://github.com/JuliaLang/julia/assets/1582097/9217805f-dc38-45b6-9cc4-44af5e219031">

NHDaly committed:
Allocation Profiler: Types for all allocations

Before this PR, we were missing the types for allocations in two cases:
1. allocations from codegen
2. allocations in gc_managed_realloc_

The second one is easy: those are always used for `buffer`s, right?

For the first one: we pass the type through to the allocation function,
so that we can log the allocation with the correct type.

NHDaly commented:
(Similar issue here, btw):
```julia
julia> using Profile
Precompiling Profile
  ✗ Profile
  0 dependencies successfully precompiled in 1 seconds
[ Info: Precompiling Profile [9abbd945-dff8-562f-b5e8-e1ebf5ef1b79]
Incorrect number of arguments passed to called function!
  %box = call noalias nonnull dereferenceable(48) {} addrspace(10)* @ijl_gc_pool_alloc_typed(i8* %34, i32 1232, i32 48, {}* %30) #22
in function jfptr_parse_flat_910
LLVM ERROR: Broken function found, compilation aborted!
```

NHDaly commented:
(Similar issue here, btw):
```julia
julia> using Profile
Precompiling Profile
  ✗ Profile
  0 dependencies successfully precompiled in 2 seconds
[ Info: Precompiling Profile [9abbd945-dff8-562f-b5e8-e1ebf5ef1b79]
Call parameter type does not match function signature!
  %"+Main.Base.Dict#12848" = load {}*, {}** @SUM.MainDOT.BaseDOT.DictYY.1284, align 8, !dbg Call parameter type does not match function signature!
  !269, !tbaa !233, !alias.scope !234, !noalias !237, !nonnull !0, !dereferenceable !248, !align !245
 i64  %newstruct = call noalias nonnull dereferenceable(80) {} addrspace(10)* @ijl_gc_pool_alloc_typed(i8* %ptls_load1314, i32% 17 = inttoptr i64 %Tuple1328 to {}*
 i64, i32 80, {}* nonnull %"+Main.B  a%sboxe = .callD icnoalias nonnull dereferenceable(32) {} addrspace(10)* t@#ijl_gc_pool_alloc_typed1(2i8848") #21, *! d%b21g,  i!32269 1184
, i32 32, {}* %17) #21
in function julia_Dict_1299
LLVM ERROR: Broken function found, compilation aborted!
in function jfptr_ht_keyindex2_shorthashNOT._1306
LLVM ERROR: Broken function found, compilation aborted!```
EDIT: oops, previous comment was mistaken.

NHDaly commented:
Aha! I fixed it with a small `builder.CreatePtrToInt(tag, T_size)`. I'm not sure why it was needed here when it wasn't needed elsewhere.. 🤔 but it makes sense I think that it should be needed.

NHDaly commented:
This PR is actually really nice and simple, and assuming I'm not missing any reason why we were setting the tag in LLVM IR, hopefully this is a nice little change that solves the profiler issue! 😊 

Thanks in advance for the review! :)

vchuravy commented on src/gc.c:
```diff
@@ -3776,7 +3793,8 @@ static void *gc_managed_realloc_(jl_ptls_t ptls, void *d, size_t sz, size_t olds
     errno = last_errno;
-    maybe_record_alloc_to_profile((jl_value_t*)b, sz, jl_gc_unknown_type_tag);
+    // gc_managed_realloc_ is currently used exclusively for resizing array buffers.
+    maybe_record_alloc_to_profile((jl_value_t*)b, sz, (jl_datatype_t*)jl_buff_tag);
```
Since this is realloc should we record the delta?

NHDaly replied:
ooh yeah great idea! 💡

NHDaly replied:
```suggestion
    maybe_record_alloc_to_profile((jl_value_t*)b, sz - oldsz, (jl_datatype_t*)jl_buff_tag);
```

Do you think we should use the supplied `sz` or the `allocsz`, computed after alignment? For pool allocated objects, we're using the `osize` not the object's actual size. So `allocsz - oldsz` seems most similar here to me?

NHDaly replied:
Do you think we should use the supplied `sz` or the `allocsz`, computed after alignment? For pool allocated objects, we're using the `osize` not the object's actual size. So `allocsz - oldsz` seems most similar here to me?

```suggestion
    maybe_record_alloc_to_profile((jl_value_t*)b, allocsz - oldsz, (jl_datatype_t*)jl_buff_tag);
```

NHDaly replied:
@gbaraldi's suggestion:
- Let's report the delta,
- and only record the alloc if the delta is positive.

NHDaly replied:
Going with this as the final change:
```suggestion
    if (allocsz > oldsz)
        maybe_record_alloc_to_profile((jl_value_t*)b, allocsz - oldsz, (jl_datatype_t*)jl_buff_tag);
    end
```

vchuravy commented:
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/50337#issuecomment-1612406715) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/a1ac997_vs_cf34aa2/report.md).

NHDaly commented:
@vchuravy: What do you make of those benchmark results? There's a surprising amount of _improvements_, which doesn't really make sense to me, and makes me worried that the results are just quite noisy.

There are more regressions than improvements, to be sure. The worst offender is this one:
```
["tuple", "reduction", ("sum", "(2,)")]        1.71 (5%) ❌	
```

NHDaly commented on src/gc.c:
```diff
@@ -1316,14 +1324,23 @@ STATIC_INLINE jl_value_t *jl_gc_pool_alloc_inner(jl_ptls_t ptls, int pool_offset
+JL_DLLEXPORT jl_value_t *jl_gc_pool_alloc_typed(jl_ptls_t ptls, int pool_offset,
+                                        int osize, jl_value_t* type)
+{
+    jl_value_t *val = jl_gc_pool_alloc_inner(ptls, pool_offset, osize);
+    jl_set_typeof(val, type);
+    maybe_record_alloc_to_profile(val, osize, (jl_datatype_t*)type);
```
The other variant on this PR could be that we _only_ set record the alloc here, but don't actually call set_typeof, and keep the tag setting back in the generated code.

......... Do you think that would be a better change? It would keep things closer to how they were before this, and i don't see how it could be any _worse_, performance-wise. And it might be possible to do better than this generic C function if you know the context?

NHDaly replied:
On the other hand, it means compiling that instruction over and over, which means longer compile times. And if it actually makes no difference, then factoring it out into this C function would make compilation slightly cheaper.

NHDaly replied:
K, i pushed that up as another variant for us to consider, in 45f3f5ea9b.

vchuravy replied:
I think I liked the previous version better, the reasoning being that it's kina weird to write at a negative offer of an object. Iirc we promise a certain amount of valid bytes and then we suddenly write to the bytes in front of it.

NHDaly replied:
I think I liked the change better too, but I like the idea of keeping the changes minimal I guess.

We're talking about it with @gbaraldi and @jpsamaroo in the Profiling call now - and Gabriel points out that there is potentially more optimization opportunity by exposing more to LLVM.

For example, if someone tries to read the type afterwards, can it elide the read? ... Though maybe that isn't even possible anyway since this would be volatile memory...

gbaraldi replied:
We are already load and store the type tag a bunch from JITted and maybe this allows for an optimization.

vchuravy replied:
So the optimization that could happen is store-load forwarding. Iff LLVM can prove that no-one else could modify the value pointed to in-between (I do not see how it could prove that).

We already do tag forwarding.

NHDaly replied:
yeah, that's my feeling too Valentin. I don't know how it could prove that.

But i do like the idea of keeping this change more limited in scope.. :/

NHDaly replied:
But maybe that's the coward's answer, and I should Be Bold and make the change!? Dunno.

It does seem a bit wrong to be changing a fundamental part of codegen as part of a profiling PR though.....

Maybe we can start with this way, and then follow up with another PR that moves the store into the C function?

vchuravy replied:
You are already modifying codegen.

But maybe @keno or @vtjnash know why we do the store in emitted IR.

NHDaly replied:
Thanks for the review, @vtjnash. Can you chime in on the above question? Do you know why we do the store in the emitted IR, rather than in C? The existing comments imply that this was intentional, and it'd be great to know.

(I'll probably merge the PR this way, after adjusting the docs, and then open a separate PR to move it into C instead?)

vtjnash replied:
I think we used to use that for some type-tag load-store forwarding, but we do that differently now

NHDaly committed:
Switch back to setting the type tag in LLVM IR, rather than in C.

The tradeoff here is:
- the compiler gets to see the set instruction and it can participate in optimization
- but the code size and compilation time of every allocation increases by 1 instruction.

vchuravy commented:
> @vchuravy: What do you make of those benchmark results? There's a surprising amount of _improvements_, which doesn't really make sense to me, and makes me worried that the results are just quite noisy.
> 
> There are more regressions than improvements, to be sure. The worst offender is this one:
> 
> ```
> ["tuple", "reduction", ("sum", "(2,)")]        1.71 (5%) ❌	
> ```

It's pretty much just noise

vtjnash commented:
@nanosoldier `runbenchmarks("find", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/50337#issuecomment-1615099962) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/638a888_vs_196a5b4/report.md).

vchuravy commented:
You will also need to adjust the docs of the allocation profiler.

NHDaly commented:
And fix the generated code tests, once we decide what we want the final codegen behavior to be.

NHDaly committed:
Review suggestion: report (positive) delta in realloc

NHDaly committed:
Update documentation for the Allocation Profiler; now we have all types

Add docs section on external profile visualizer tools

NHDaly committed:
First attempt at fixing llvm tests...

I frankly don't know enough to do this yet, I think.

vchuravy commented on doc/src/manual/profile.md:
```diff
@@ -338,15 +338,77 @@ argument can be passed to speed it up by making it skip some allocations.
+!!! compat "Julia 1.11"
+
+    Older versions of Julia could not capture types in all cases. In older versions of
+    julia, if you see an allocation of type `Profile.Allocs.UnknownType`, it means that
```
```suggestion
    Julia, if you see an allocation of type `Profile.Allocs.UnknownType`, it means that
```

vchuravy commented on doc/src/manual/profile.md:
```diff
@@ -338,15 +338,77 @@ argument can be passed to speed it up by making it skip some allocations.
+    the allocation was coming from generated code produced by the compiler. See
+    [issue #43688](https://github.com/JuliaLang/julia/issues/43688) for more info.
+
+    In Julia 1.11+, all allocations should have a type reported.
```
```suggestion
    Since Julia 1.11, all allocations should have a type reported.
```

vchuravy commented on doc/src/manual/profile.md:
```diff
@@ -338,15 +338,77 @@ argument can be passed to speed it up by making it skip some allocations.
+
+Here is an example of how to invoke the Allocation profiler. A good number of samples to aim
+for is around 1 - 10 thousand. Too many, and the profile visualizer can get overwhelmed, and
+profiling will be slow. Too few, and you don't have a representative sample.
```
```suggestion
profiling will be slow. Too few, and you don't have a representative sample.

```

NHDaly committed:
Docs suggestions

Co-authored-by: Valentin Churavy <vchuravy@users.noreply.github.com>

ojwoodford commented on doc/src/manual/profile.md:
```diff
@@ -338,15 +338,78 @@ argument can be passed to speed it up by making it skip some allocations.
+       end
+500000
+
+julia> prof = Profile.Allocs.fetch();
```
This is not required, right?

Most people will just copy the example. So in my view the example should be the simplest thing possible.

Something like:

```
using Profile, PProf
Profile.Allocs.clear()
Profile.Allocs.@profile sample_rate=0.001 my_function()
PProf.Allocs.pprof()
```

NHDaly replied:
Thanks for the review. Thanks that was excellent feedback! :)

I've added this very simple example to the doc as the first example, and then i explain more about why i was going into the details in the second one. 👍 Let me know if that looks good? (We can also do follow-up PRs if you want to add more!

vtjnash commented on test/llvmpasses/alloc-opt-gcframe.ll:
```diff
@@ -14,17 +14,18 @@ target datalayout = "e-m:o-i64:64-f80:128-n8:16:32:64-S128"
 ; CHECK-NOT: @julia.gc_alloc_obj
 
```
https://llvm.org/docs/TestingGuide.html#generating-assertions-in-regression-tests

NHDaly committed:
Now i know about the LLVM tests! :tada:

NHDaly committed:
Fix rest of the llvmpasses tests

NHDaly committed:
typo

NHDaly committed:
Fix jl_gc_alloc_typed from codegen.

This must never happen in any of the tests, other than this one llvm
unit test! :o

NHDaly committed:
PR Feedback: Add simpler example, and explain the in-depth example more.

NHDaly commented:
Hokay, i think that should be the last issue with the llvmpasses tests! 🤞 Thanks @topolarity! 😊

I'll check back in soon and if they're all green we should be good to merge. The other test failure is the same as what's broken on master: https://github.com/JuliaLang/julia/issues/50550

NHDaly commented:
😁 i never actually tried it end-to-end in PProf! Now we have all the types! 🙌 

Here are some examples:
<img width="887" alt="Screenshot 2023-07-29 at 8 48 16 PM" src="https://github.com/JuliaLang/julia/assets/1582097/9217805f-dc38-45b6-9cc4-44af5e219031">

<img width="926" alt="Screenshot 2023-07-29 at 10 30 02 PM" src="https://github.com/JuliaLang/julia/assets/1582097/f33db3bd-992d-4aba-8d03-1e0862843503">

NHDaly commented:
Do you think this ASAN error is real? Or noise?
```
JIT session error: Duplicate section
JIT session error: Duplicate section
JIT session error: Duplicate section
JIT session error: Duplicate section
JIT session error: Duplicate section
JIT session error: Duplicate section
JIT session error: Duplicate section
JIT session error: Duplicate section
JIT session error: Duplicate section
JIT session error: Duplicate section
JIT session error: Duplicate section
```
https://buildkite.com/julialang/julia-master/builds/26289#0189a440-49b4-4569-a25f-3b10c57deeec

NHDaly committed:
Update profile.md

NHDaly committed:
Update profile.md

NHDaly commented:
I think this is ready to merge. There was the failure from https://github.com/JuliaLang/julia/issues/50550, and then a few failures in the Allow Fail section. There was a [segfault](https://buildkite.com/julialang/julia-master/builds/26291#0189a4bf-a591-4b5a-bf4f-8bf9b19daf90) in the `:linux: test i686-linux-gnu`, but it's in the allow-fail, so it's maybe fine / expected?

I'm going to mark this as `Merge Me`, but i'd love if someone can confirm that i'm understanding the failures correctly. Thanks! 😊

NHDaly commented:
Thank you! :)
